lib/pure/parsejson.nim

```diff
@@ -317,35 +327,98 @@ proc skip(my: var JsonParser) =
       break
   my.bufpos = pos
 
-proc parseNumber(my: var JsonParser) =
-  var pos = my.bufpos
-  if my.buf[pos] == '-':
-    add(my.a, '-')
-    inc(pos)
-  if my.buf[pos] == '.':
-    add(my.a, "0.")
-    inc(pos)
+template jsOrVmBlock(caseJsOrVm, caseElse: untyped): untyped =
+  when nimvm:
+    block:
+      caseJsOrVm
   else:
-    while my.buf[pos] in Digits:
-      add(my.a, my.buf[pos])
-      inc(pos)
-    if my.buf[pos] == '.':
-      add(my.a, '.')
-      inc(pos)
-  # digits after the dot:
-  while my.buf[pos] in Digits:
-    add(my.a, my.buf[pos])
-    inc(pos)
-  if my.buf[pos] in {'E', 'e'}:
-    add(my.a, my.buf[pos])
-    inc(pos)
-    if my.buf[pos] in {'+', '-'}:
-      add(my.a, my.buf[pos])
-      inc(pos)
-    while my.buf[pos] in Digits:
-      add(my.a, my.buf[pos])
-      inc(pos)
-  my.bufpos = pos
+    block:
+      when defined(js) or defined(nimscript):
+        # nimscript has to be here to avoid semantic checking of caseElse
+        caseJsOrVm
+      else:
+        caseElse
+
+template doCopy(a, b, start, endp1: untyped): untyped =
+  jsOrVmBlock:
+    a = b[start ..< endp1]
+  do:
+    let n = endp1 - start
+    if n > 0:
+      a.setLen n
+      copyMem a[0].addr, b[start].addr, n
+
+proc i64(c: char): int64 {.inline.} = int64(ord(c) - ord('0'))
+
+proc pow10(e: int64): float {.inline.} =
+  const p10 = [1e-22, 1e-21, 1e-20, 1e-19, 1e-18, 1e-17, 1e-16, 1e-15, 1e-14,
+               1e-13, 1e-12, 1e-11, 1e-10, 1e-09, 1e-08, 1e-07, 1e-06, 1e-05,
+               1e-4, 1e-3, 1e-2, 1e-1, 1.0, 1e1, 1e2, 1e3, 1e4, 1e5, 1e6, 1e7,
+               1e8, 1e9]                        # 4*64B cache lines = 32 slots
+  if -22 <= e and e <= 9:
+    return p10[e + 22]                          # common case=small table lookup
+  result = 1.0
+  var base = 10.0
+  var e = e
+  if e < 0:
+    e = -e
+    base = 0.1
+  while e != 0:
+    if (e and 1) != 0:
+      result *= base
+    e = e shr 1
+    base *= base
+
+proc parseNumber(my: var JsonParser): TokKind {.inline.} =
```

Comment:
is not the proc kinda too big to be `{.inline.}` ?. :thinking:

Reply:
No.  In the by far most common case it is just one line (two almost always perfectly predicted compare-jumps and one array reference).  

It is only used just once, for human mental abstraction.  So, there is no duplicated generated code.

Also, people are too sparing of `{.inline.}` in my view.  Function call overhead is like 12..15 cycles * 3..6way superscalar =~ 36..90 *dynamic* instruction slots.  So, if duplication is not a concern, that is the the rough amount of work to compare to.

If you were really going to browbeat me over inline then you should complaing about `parseNumber` being inline, not `pow10`. ;-)  But many numbers to parse might be just single-digits which take only a few cycles.  In fact, the more numbers you have the more likely many are single digits.  So, the function call overhead would then be quite large compared to the work done.

Reply:
Another thing inline buys you here is putting `parseNumber` C code in the same translation unit as the parsing itself.  So, the C compilers metrics/heuristics for inlining can at least be applied.  Otherwise you rely on LTO which a lot of people do not enable.  Given that this whole bit of work was to get really fast speed, it seems appropriate for that reason as well as the "many single digit" cases.  If you do PGO builds the compiler will actually try to *measure* if inlining is good or not.  According to the GCC guys that is the main benefit of PGO.  I am always getting 1.5x to 2.0x speed ups from PGO with Nim which suggests to me that our `{.inline.}` choices are not generally so great (or at least a few critical ones...).

Reply:
Oh, and also `parseNumber` is also only called in one place in `getTok`.  So, still no extra L1 I-cache style resource hit, and I also added `{.inline.}` there to get the meat of all this logic into the translation unit of whatever is using the json parser.

Reply:
Also, Nim's `{.inline.}` just marks it in the generated C as `inline`.  So, it's still up the C/C++ compiler code metrics what ultimately happens. Those surely do take into account the statically observable number of expansions/substitutions and their sizes.  So, the many calls to `getTok` from `json` are not a real problem.  But if anyone really wants me to remove them I can.  I always use (and recommend using) PGO when performance matters, myself.